Improve error stack

ISSOtm opened the pull request:
The old error stack was fairly obtuse and hard to use for debugging.
This improves it notably by ensuring all line numbers are relative to the file, and not, say, the macro definition.
This is a breaking change if you were parsing the old stack, but the change should be painless, and the new stack only brings more info.
The syntax is unchanged for files, macros see their name prefixed with the file they're defined in and a pair of colors, REPT blocks simply append a '::REPT~n' to the context they're in, where 'n' is the number of iterations the REPT has done.
This is especially helpful in macro-heavy code such as [rgbds-structs](https://github.com/ISOtm/rgbds-structs).

ISSOtm commented:
If accepted, resolves #392

ISSOtm commented:
The code I was using to test the error stack works, but I managed to miss some cases, it seems. Will check further.

AntonioND commented:
Avoid having merge commits in pull requests.

ISSOtm commented:
I know; "Merge and squash" will get rid of that.

dbrotz commented:
I think it would be better if the current rept iteration could be formatted using sprintf each time instead of using that logic to increment it. That way the code would be easier to follow.

ISSOtm commented:
Rebased against master manually, which got rid of the merge commit.

ISSOtm commented:
Changed the REPT count updating code to use `strtoul` and `snprintf` instead of manipulating the characters directly, and fixed error output in case the new count cannot be written (see fstack.c:139)

dbrotz commented on src/asm/fstack.c:
```diff
@@ -115,6 +124,29 @@ static int32_t popcontext(void)
+			nNbCharsWritten = snprintf(pREPTIterationWritePtr,
+						   nNbCharsLeft, "%lu",
+						   nREPTIterationNo + 1);
+			if (nNbCharsWritten > nNbCharsLeft) {
```
This should be `>=` instead of `>`. The return value doesn't count the null terminator, but the value passed to `snprintf()` (`nNbCharsLeft`) does count it.

ISSOtm replied:
> The functions snprintf() and vsnprintf() do not write  more  than  size
>        bytes  (including the terminating null byte ('\0')).  If the output was
>        truncated due to this limit, then the return value  is  the  number  of
>        characters  (excluding the terminating null byte) which would have been
>        written to the final string if enough space had been available.   **Thus,
>        a  return  value  of  size or more means that the output was truncated.**

Looks like I misread, oops.

dbrotz commented on src/asm/fstack.c:
```diff
@@ -115,6 +124,29 @@ static int32_t popcontext(void)
+						   nREPTIterationNo + 1);
+			if (nNbCharsWritten > nNbCharsLeft) {
+				/*
+				 * The string is not not terminated anymore,
```
`snprintf()` always appends a null terminator, so the string will be terminated. I think this should say: "The string is truncated,".

ISSOtm replied:
Reading `snprintf`'s manpage, I see no mention of that, only that "at most `size` bytes will be written". Worst case, this can be changed to "The string hasn't been properly written" to be more catch-all.

ISSOtm replied:
I've changed the first line to "The string is probably corrupted somehow," which should be correct and catch-all enough.

dbrotz replied:
That's fine. According to the C11 standard, `snprintf()` has to write the null terminator as long as `n` is not zero. From 7.21.6.5/2:
> If **n** is zero, nothing is written, and **s** may  be  a  null  pointer.   Otherwise,  output  characters  beyond  the **n-1**st  are discarded rather than being written to the array, and a null character is written at the end of  the  characters  actually  written  into  the  array.

You're right that Linux man pages sound a bit confusing.
> The functions snprintf() and vsnprintf() write at most size bytes (including the terminating null byte ('\0')) to str.

I am pretty sure they are trying to say the same thing as the standard, but they didn't express it very well.

ISSOtm replied:
"Including **a** terminating null byte" would probably be clearer.

ISSOtm commented:
Since I merged simple fixes to @dbrotz's two complaints, I'll consider this approved; merging.
